File: positive-reputation-feed-py/src/findings.py
# Copyright 2022 The Forta Foundation

from forta_agent import Finding, FindingType, FindingSeverity, EntityType


class PositiveReputationFinding:

    @staticmethod
    def create_finding(address: str, base_bots: list) -> Finding:
        base_bots_metadata = {"base_bots_" + str(i): address for i, address in enumerate(base_bots, 1)}

        confidence = 0.5 if len(base_bots) > 1 else 0.3

        labels = [{"entity": address,
                   "entity_type": EntityType.Address,
                   "label": "attacker",
                   "confidence": confidence}]  # very low

        return Finding({
                       'name': 'Positive Reputation Assigned',
                       'description': f'EOA {address} was assigned positive reputation.',
                       'alert_id': 'POSITIVE-REPUTATION-1',
                       'type': FindingType.Info,
                       'severity': FindingSeverity.Info,
                       'metadata': base_bots_metadata,
                       'labels': labels
                       })

File: chain-patrol-ts/src/findings.ts
import { Finding, FindingSeverity, FindingType, Label, EntityType } from "forta-agent";
import { utils } from "ethers";
import { UnalertedAsset } from "./types";

export function createBlockedAssetFinding(asset: UnalertedAsset): Finding {
  const { content, type, status, updatedAt, reason, reportId, reportUrl }: UnalertedAsset = asset;

  const resultString: string = content + updatedAt;
  const uniqueKey: string = utils.keccak256(utils.toUtf8Bytes(resultString));

  return Finding.fromObject({
    name: "A scam has been detected by ChainPatrol",
    description: `ChainPatrol detected scam: ${content}`,
    alertId: "CHAINPATROL-SCAM-ASSET",
    severity: FindingSeverity.Critical,
    type: FindingType.Scam,
    uniqueKey,
    protocol: "N/A",
    metadata: {
      type,
      status,
      updatedAt,
      reason: reason ? reason : "",
      reportId: reportId ? reportId.toString() : "",
      reportUrl: reportUrl ? reportUrl : "",
    },
    labels: [
      Label.fromObject({
        entity: content,
        entityType: EntityType.Url,
        label: `Blocked ${type}`,
        confidence: 0.99,
        remove: false,
        metadata: {
          type,
          status,
          updatedAt,
          reason: reason ? reason : "",
          reportId: reportId ? reportId.toString() : "",
          reportUrl: reportUrl ? reportUrl : "",
        },
      }),
    ],
  });
}
